### PR TITLE
Split client and provider into separate packages

### PR DESCRIPTION
**Pull request by palson-cf**
## Overview

Create **client** package:
- `client.go`:
  - `Client` struct;
  - `NewClient()`;
- `pipeline.go`:
  - `Pipeline` struct;
  - CRUD methods;
- `project.go`:
  - `Project` struct;
  - CRUD methods;

### CODE REVIEW

File: client/pipeline.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+type RuntimeEnvironment struct {
+	Name        string `json:"name,omitempty"`
+	Memory      string `json:"memory,omitempty"`
+	CPU         string `json:"cpu,omitempty"`
+	DindStorage string `json:"dindStorage,omitempty"`
+}
+
+func (t *Trigger) SetVariables(variables map[string]interface{}) {
```
Comment (sharon-codefresh) [inline client/pipeline.go:58]: why we use 
map[string]interface{}
instead of:
map[string]string ?

File: client/utils.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package client
+
+// Variable spec
+type Variable struct {
```
Comment (sharon-codefresh) [inline client/utils.go:4]: maybe we should rename this to PipelineVariable?

File: codefresh/provider.go [hunk 3]
```diff
@@ -27,10 +35,9 @@ func Provider() *schema.Provider {
...
 
 func configureProvider(d *schema.ResourceData) (interface{}, error) {
-	config := Config{
-		APIServer: d.Get("api_url").(string),
-		Token:     d.Get("token").(string),
-	}
 
-	return &config, nil
+	apiURL := d.Get("api_url").(string)
+	token := d.Get("token").(string)
```
Comment (sharon-codefresh) [inline codefresh/provider.go:40]: why we use "d" as variable name?

File: codefresh/resource_project.go [hunk 4]
```diff
@@ -57,67 +38,68 @@ func resourceProject() *schema.Resource {
 }
 
 func resourceProjectCreate(d *schema.ResourceData, meta interface{}) error {
```
Comment (sharon-codefresh) [inline codefresh/resource_project.go:40]: why we use "d" for variable?

File: codefresh/resource_pipeline.go [hunk 5]
```diff
@@ -291,93 +282,151 @@ func mapPipelineToResource(cfObject codefreshObject, d *schema.ResourceData) err
...
+		pipeline.Spec.SpecTemplate = &cfClient.SpecTemplate{
+			Location: d.Get("spec.0.spec_template.0.location").(string),
+			Repo:     d.Get("spec.0.spec_template.0.repo").(string),
+			Path:     d.Get("spec.0.spec_template.0.path").(string),
+			Revision: d.Get("spec.0.spec_template.0.revision").(string),
+			Context:  d.Get("spec.0.spec_template.0.context").(string),
+		}
+	}
+
+	if _, ok := d.GetOk("spec.0.runtime_environment"); ok {
```
Comment (sharon-codefresh) [inline codefresh/resource_pipeline.go:400]: what is "ok" why not "err"?

File: client/utils.go (hunk 2 above)
Comment (palson-cf) [reply client/utils.go:4]: it is used by both pipeline and project
```
	Variables   []Variable `json:"variables,omitempty"`
```

File: codefresh/provider.go (hunk 3 above)
Comment (palson-cf) [reply codefresh/provider.go:40]: it is a common pattern for `*schema.ResourceData` struct
https://www.terraform.io/docs/extend/writing-custom-providers.html#implement-create

File: codefresh/resource_pipeline.go (hunk 5 above)
Comment (palson-cf) [reply codefresh/resource_pipeline.go:400]:
```
// GetOk returns the data for the given key and whether or not the key
// has been set to a non-zero value at some point.
//
// The first result will not necessarilly be nil if the value doesn't exist.
// The second result should be checked to determine this information.
func (d *ResourceData) GetOk(key string) (interface{}, bool) {
	r := d.getRaw(key, getSourceSet)
	exists := r.Exists && !r.Computed
	if exists {
		// If it exists, we also want to verify it is not the zero-value.
		value := r.Value
		zero := r.Schema.Type.Zero()

		if eq, ok := value.(Equal); ok {
			exists = !eq.Equal(zero)
		} else {
			exists = !reflect.DeepEqual(value, zero)
		}
	}

	return r.Value, exists
}
```

I think `ok` is a better name for this var.
`if <ok=this parameter is set> { do something }`

**Pull request merged**